components/webext-storage/src/db.rs

```diff
@@ -0,0 +1,270 @@
+}
+
+/// The entry-point to getting a database connection. No enforcement of this
+/// as a singleton is made - that's up to the caller. If you make multiple
+/// StorageDbs pointing at the same physical database, you are going to have a
+/// bad time.
+pub struct StorageDb {
+    db_path: PathBuf,
+    pub writer: Arc<Mutex<Connection>>,
+}
+impl StorageDb {
+    /// Create a new, or fetch an already open, StorageDb backed by a file on disk.
+    pub fn new(db_path: impl AsRef<Path>) -> Result<Arc<Self>> {
+        let db_path = normalize_path(db_path)?;
+        Self::new_or_existing(db_path)
+    }
+
+    /// Create a new, or fetch an already open, memory-based StorageDb. You must
+    /// provide a name, but you are still able to have a single writer and many
+    ///  reader connections to the same memory DB open.
+    pub fn new_memory(db_path: &str) -> Result<Arc<Self>> {
+        let name = PathBuf::from(format!("file:{}?mode=memory&cache=shared", db_path));
+        Self::new_or_existing(name)
+    }
+
+    fn new_or_existing(db_path: PathBuf) -> Result<Arc<Self>> {
+        // We always create the read-write connection for an initial open so
+        // we can create the schema and/or do version upgrades.
+        let conn = Connection::open_with_flags(db_path.clone(), get_write_flags())?;
+        match init_sql_connection(&conn, true) {
+            Ok(()) => Ok(Arc::new(Self {
+                db_path,
+                writer: Arc::new(Mutex::new(conn)),
+            })),
+            Err(e) => {
+                // like with places, failure to upgrade means "you lose your data"
+                if let ErrorKind::DatabaseUpgradeError = e.kind() {
+                    fs::remove_file(&db_path)?;
+                    Self::new_or_existing(db_path)
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
+
+    /// Open a read-only connection to the database.
+    pub fn open_read_connection(&self) -> Result<Connection> {
+        let conn = Connection::open_with_flags(self.db_path.clone(), get_read_flags())?;
+        init_sql_connection(&conn, false)?;
+        Ok(conn)
+    }
+
+    /// Close a connection to the database. If the connection is the write
+    /// connection, you will have closed the only connection able to write, so
+    /// don't do that unless you are sure.
+    pub fn close_connection(&self, _conn: Connection) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn init_sql_connection(conn: &Connection, is_writable: bool) -> Result<()> {
+    let initial_pragmas = "
+        -- Cargo-culted from places.
+        PRAGMA page_size = 32768;
+
+        -- Disable calling mlock/munlock for every malloc/free.
+        -- In practice this results in a massive speedup, especially
+        -- for insert-heavy workloads.
+        -- XXX - is this relevant?
+        PRAGMA cipher_memory_security = false;
```

Comment:
I don't think this should be relevant unless we're encrypting this database, should it? /cc @thomcc

Reply:
:) I think schema management, and maybe even db management in general, is somewhere that remerge-v2 might be able to help with.